### PR TITLE
feat: notify on workspace creation

### PR DESCRIPTION
**Pull request by DanielleMaywood**
Relates to https://github.com/coder/coder/issues/15845

This PR introduces notifications for when workspaces are created.

Each notification contains some metadata about the workspace that has been created. I've intentionally limited this information to just IDs and names, as we can always add more data if requested/desired.

Adding the new notification caused other unrelated tests to fail due to how they were checking for the right notification. I've updated these tests to use a new helper function `SentWithTemplateID` to hopefully reduce the likelihood of further notifications to cause a cascade of test failures.

### CODE REVIEW

File: coderd/workspaces_test.go [hunk 1]
```diff
@@ -571,6 +571,29 @@ func TestPostWorkspacesByOrganization(t *testing.T) {
 		require.Equal(t, http.StatusConflict, apiErr.StatusCode())
 	})
 
+	t.Run("CreateSendsNotification", func(t *testing.T) {
+		t.Parallel()
+
+		enqueuer := notificationstest.FakeEnqueuer{}
+		client := coderdtest.New(t, &coderdtest.Options{IncludeProvisionerDaemon: true, NotificationsEnqueuer: &enqueuer})
+		user := coderdtest.CreateFirstUser(t, client)
```
Comment (johnstcn) [inline coderd/workspaces_test.go:579]:
```suggestion
		user := coderdtest.CreateFirstUser(t, client)
		memberClient, memberUser := coderdtest.CreateAnotherUser(t, client, user.OrganizationID)
```

File: coderd/workspaces_test.go [hunk 2]
```diff
@@ -571,6 +571,29 @@ func TestPostWorkspacesByOrganization(t *testing.T) {
...
+
+		version := coderdtest.CreateTemplateVersion(t, client, user.OrganizationID, nil)
+		template := coderdtest.CreateTemplate(t, client, user.OrganizationID, version.ID)
+		coderdtest.AwaitTemplateVersionJobCompleted(t, client, version.ID)
+
+		workspace := coderdtest.CreateWorkspace(t, client, template.ID)
+		coderdtest.AwaitWorkspaceBuildJobCompleted(t, client, workspace.LatestBuild.ID)
```
Comment (johnstcn) [inline coderd/workspaces_test.go:586]:
```suggestion
		workspace := coderdtest.CreateWorkspace(t, memberClient, template.ID)
		coderdtest.AwaitWorkspaceBuildJobCompleted(t, memberClient, workspace.LatestBuild.ID)
```

File: coderd/workspaces_test.go [hunk 3]
```diff
@@ -571,6 +571,29 @@ func TestPostWorkspacesByOrganization(t *testing.T) {
...
+
+		sent := enqueuer.SentWithTemplateID(notifications.TemplateWorkspaceCreated)
+		require.Len(t, sent, 1)
+		require.Equal(t, user.UserID, sent[0].UserID)
```
Comment (johnstcn) [inline coderd/workspaces_test.go:590]:
```suggestion
		require.Equal(t, memberUser.ID, sent[0].UserID)
```

File: coderd/database/migrations/000279_workspace_create_notification.up.sql
Comment (johnstcn) [inline coderd/database/migrations/000279_workspace_create_notification.up.sql:1]: Obligatory prod to double-check migration number before merging :-)

File: coderd/notifications/notificationstest/fake_enqueuer.go [hunk 4]
```diff
@@ -97,3 +97,16 @@ func (f *FakeEnqueuer) Sent() []*FakeNotification {
...
+	defer f.mu.Unlock()
+
+	sent := []*FakeNotification{}
+	for _, notif := range f.sent {
+		if notif.TemplateID == id {
+			sent = append(sent, notif)
+		}
+	}
+	return sent
+}
```
Comment (johnstcn) [inline coderd/notifications/notificationstest/fake_enqueuer.go:112]: suggestion: we could alternatively modify the signature of `FakeEnqueuer.Sent()` to accept an arbitrary number of functions to match notifications we want:

```
func (f *FakeEnqueuer) Sent(matchers ...func(*FakeNotification) bool) []*FakeNotification {
  ...
}
```

File: coderd/workspaces_test.go
Comment (johnstcn) [inline coderd/workspaces_test.go:1]: Can we also add a test that ensures that when someone with the ability to create workspaces on behalf of other users creates a workspace for a different user, the notification is sent to the owner of the workspace?

**Review by johnstcn — APPROVED** [review]
LGTM, some suggestions below but nothing blocking:

- Regarding your `coderd` tests, it's good practice to "drop privileges" by creating a separate user with the minimum permission set for the behaviour you want to test, and testing with that user. This will catch RBAC-related bugs.
- As it's also possible to create workspaces on behalf of other users, we should also test the behaviour in this workflow.

File: coderd/notifications/notificationstest/fake_enqueuer.go (hunk 4 above)
Comment (DanielleMaywood) [reply coderd/notifications/notificationstest/fake_enqueuer.go]: Have now done so with a matcher for template ids

File: coderd/notifications/notificationstest/fake_enqueuer.go [hunk 5]
```diff
@@ -92,8 +92,31 @@ func (f *FakeEnqueuer) Clear() {
...
+		}
+	}
+
+	return sent
+}
+
+func WithTemplateID(id uuid.UUID) func(*FakeNotification) bool {
+	return func(n *FakeNotification) bool {
+		return n.TemplateID == id
+	}
```
Comment (johnstcn) [inline coderd/notifications/notificationstest/fake_enqueuer.go:121]: Nice :-)

**Review by johnstcn — APPROVED** [review]

**Pull request merged**